### PR TITLE
[nrfconnect] Added scripts to generate factory data partition

### PR DESCRIPTION
**Pull request by ArekBalysNordic**
#### Problem
The nRFconnect platform should contain a script to generate factory data.

#### Change overview

- Added two scripts to generate factory data partition for nRFconnect.
- The first script generates a JSON file based on provided arguments.
- The second script takes a JSON file as an argument 
to create a CBOR file. A CBOR file may be saved to .hex and .bin files.
Then ready factory data can be flashed into the flash memory of the nRFconnect device.
- Added a JSON schema to validate factory data JSON object.

#### Testing

It was tested locally on a Linux machine.

### CODE REVIEW

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        "spake2_salt",
+        "spake2_verifier"
+    ],
+    "properties": {
+        "sn": {
+            "description": "Serial number of device",
+            "type": "string",
+            "maxLength": 32
+        },
+        "date": {
```
Comment (markaj-nordic) [inline scripts/tools/nrfconnect/nrfconnect_factory_data.schema:26]: json shema has a dedicated type for date/time. Maybe worth checking (i.e. to get rid of regex and lengths):
https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        },
+        "dac_key": {
+            "description": "DAC Private Key in hex-string format",
+            "type": "string"
+        },
+        "pai_cert": {
+            "description": "PAI certificate in hex-string format",
+            "type": "string"
+        },
+        "cd": {
```
Comment (markaj-nordic) [inline scripts/tools/nrfconnect/nrfconnect_factory_data.schema:63]: nit: cd does not look much verbose :) decl_cert?

File: scripts/tools/nrfconnect/nrfconnect_generate_partition.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    def __init__(self, offset: int, length: int, input: str, output: str) -> None:
+        self.__ih = IntelHex()
+        self.__length = length
+        self.__offset = offset
+        self.__data_ready = False
+        self.__output = output
+        self.__input = input
+
+        self.__data_to_save = self.__convert_to_dict(self.__load_json())
```
Comment (markaj-nordic) [inline scripts/tools/nrfconnect/nrfconnect_generate_partition.py:47]: __load_json -> the same as crate_hex, raises?

**Review by markaj-nordic — COMMENTED** [review]
Some high level comments

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema (hunk 1 above)
Comment (ArekBalysNordic) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:26]: JSON-schema provides a `date` format in ISO 8601 (YYYY-MM-DD), but in specs, we have a format YYYYMMDD, so without '-'. So, I think we should be coherent with spec.

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema (hunk 2 above)
Comment (ArekBalysNordic) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:63]: I tried to reduce the names of the keys written to flash, and I wanted to be coherent with names everywhere. I see what you're worried about, and I included descriptions in the schema and the script.

**Review by jnsgsr — CHANGES_REQUESTED** [review]

File: scripts/tools/nrfconnect/generate_nrfconnect_chip_factory_data.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+                                     help="[int] Provide BLE pairing discriminator.")
+    # optional keys
+    optional_arguments.add_argument("--rd_uid", type=str,
+                                    help="[base64 string] Provide the rotating device unique ID. To generate the new rotate device unique ID use --rd_uid_gen/")
+    optional_arguments.add_argument("--passcode", type=allow_any_int,
+                                    help="[int | hex] Default PASE session passcode. (This is mandatory to generate Spake2 Verifier).")
+    optional_arguments.add_argument("--spake2p_path", type=str,
+                                    help="[string] Provide a path to spake2p. By default You can find spake2p in connectedhomeip/src/tools/spake2p directory and build it there.")
+    optional_arguments.add_argument("--spake2_verifier", type=str,
+                                    help="[base64 string] Provide Spake2 Verifier without generating it.")
```
[the remark this answers is not in the document]
Comment (ArekBalysNordic) [reply scripts/tools/nrfconnect/generate_nrfconnect_chip_factory_data.py:242]: This value is stored with `hex:` prefix too. Look at lines 107 and 109.

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            "description": "A string representation of hardware version",
+            "type": "string",
+            "minLength": 1,
+            "maxLength": 64
+        },
+        "rd_uid": {
+            "description": "A randomly-generated 128-bit or longer octet string. Length has been expanded with 'hex:' prefix",
+            "type": "string",
+            "minLength": 20,
+            "maxLength": 36
```
[the remark this answers is not in the document]
Comment (ArekBalysNordic) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:49]: You're right. It is a randomly-generated 128-bit or longer octet string, so it doesn't have a maximum length. I use 32 bytes to ensure that we have control over providing data by users.

**Review by markaj-nordic — APPROVED** [review]

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema (hunk 5 above)
[the remark this answers is not in the document]
Comment (jnsgsr) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:49]: OK

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        },
+        "cd": {
+            "description": "Certificate Declaration in hex-string format",
+            "type": "string"
+        },
+        "passcode": {
+            "description": "A default PASE session passcode",
+            "type": "integer",
+            "minimum": 1,
+            "maximum": 99999998
```
[the remark this answers is not in the document]
Comment (jnsgsr) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:71]: I took the wrong item from the specification with this comment, you are right

**Review by jnsgsr — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]

**Review by bauerschwan — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**